### PR TITLE
[CALCITE-3779] Implement BITAND, BITOR, BITXOR scalar functions

### PR DESCRIPTION
**Pull request by normanj-bitquill**
* All three functions accept either a pair of Integer or Binary values as arguments
* Binary arguments must be of the same length
* Returns NULL if any argument is NULL
* Returns a value of the same type as the first argument if both arguments are not null

### CODE REVIEW

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java [hunk 1]
```diff
@@ -15470,6 +15470,111 @@ private static void checkLogicalOrFunc(SqlOperatorFixture f) {
...
+        "Different length for bitwise operands: the first: 2, the second: 1",
+        true);
+  }
+
+  @Test void testBitOrScalarFunc() {
+    final SqlOperatorFixture f = fixture();
+    f.setFor(SqlStdOperatorTable.BITOR, VmName.EXPAND);
+    f.checkFails("bitor(^*^)", "Unknown identifier '\\*'", false);
+    f.checkScalar("bitor(2, 4)", "6", "INTEGER");
+    f.checkType("bitor(CAST(2 AS TINYINT), CAST(6 AS TINYINT))", "TINYINT");
```
Comment (mihaibudiu) [inline testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:15513]: Can we have a few tests with negative numbers, and a few with negative shorter integers?

**Review by mihaibudiu — APPROVED** [review]
I don't really understand why the result for TINYINT will be correct, since today Calcite incorrectly does not truncate the results for short arithmetic.

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java [hunk 2]
```diff
@@ -2829,6 +2829,20 @@ public static long bitAnd(long b0, long b1) {
...
   }
 
+  /** Bitwise function <code>BITAND</code> applied to a Long and int value.
+   *  Needed for handling NULL for the first argument.
+   */
+  public static long bitAnd(Long b0, int b1) {
+    return b0 & b1;
+  }
+
+  /** Bitwise function <code>BITAND</code> applied to a Long and int value.
```
Comment (NobiGo) [inline core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:2839]: Bitwise function <code>BITAND</code> applied to a int and Long value.

**Review by NobiGo — APPROVED** [review]

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java (hunk 1 above)
Comment (normanj-bitquill) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:15523]: Added a few more tests with negative numbers.

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java (hunk 2 above)
Comment (normanj-bitquill) [reply core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:2839]: Added some more tests with different integer types for the arguments (integer and bigint).

Needed to rework the return type detection. It will now use the largest integer type if both arguments are integer types. There may be ways of cleaning the code up further for the return type inference.

**Comment by normanj-bitquill** [comment]
> I don't really understand why the result for TINYINT will be correct, since today Calcite incorrectly does not truncate the results for short arithmetic.

@mihaibudiu I have created a custom `SqlReturnTypeInference` class. It will look at the arguments and choose the largest integer type (if both are integer types), otherwise use the type of the first argument. If both arguments are `NULL`, then just assume integer nullable return type.

Perhaps you mean when the result of the BITAND is converted from a long back to a TINYINT. If this is the case, do you have any examples where it fails to truncate the value correctly? No truncation should be needed, since the result cannot exceed the largest argument type.

**Comment by mihaibudiu** [comment]
I think this pattern is common in all arithmetic operations for integer types. Hopefully you could have reused some code.
Your runtime only supported int and long, so I suspected the results would be wrong for short.

File: core/src/main/java/org/apache/calcite/sql/type/ReturnTypes.java [hunk 3]
```diff
@@ -613,6 +613,59 @@ public static SqlCall stripSeparator(SqlCall call) {
...
+            } else if (isIntegerTypeLarger(integerReturnType, opType)) {
+              integerReturnType = opType;
+            }
+          } else {
+            allArgsInteger = false;
+            break;
+          }
+        }
+        if (integerReturnType != null && allArgsInteger) {
+          return typeFactory.createTypeWithNullability(integerReturnType, true);
```
Comment (mihaibudiu) [inline core/src/main/java/org/apache/calcite/sql/type/ReturnTypes.java:638]: why are all these types nullable?

File: core/src/main/java/org/apache/calcite/sql/type/ReturnTypes.java [hunk 4]
```diff
@@ -613,6 +613,59 @@ public static SqlCall stripSeparator(SqlCall call) {
...
+        }
+        for (RelDataType opType : opBinding.collectOperandTypes()) {
+          if (!opType.isNullable()) {
+            return typeFactory.createTypeWithNullability(opType, true);
+          }
+        }
+        return typeFactory.createTypeWithNullability(
```
Comment (mihaibudiu) [inline core/src/main/java/org/apache/calcite/sql/type/ReturnTypes.java:645]: When can this happen?

File: core/src/main/java/org/apache/calcite/sql/type/ReturnTypes.java [hunk 5]
```diff
@@ -613,6 +613,59 @@ public static SqlCall stripSeparator(SqlCall call) {
...
+
+  /**
+   * Returns true if type2 is of a larger integer type than type1.
+   *
+   * @param type1 first type
+   * @param type2 second type
+   * @return true if type2 is of a larger integer type than type1
+   */
+  private static boolean isIntegerTypeLarger(RelDataType type1, RelDataType type2) {
+    if (SqlTypeName.TINYINT == type1.getSqlTypeName()) {
```
Comment (mihaibudiu) [inline core/src/main/java/org/apache/calcite/sql/type/ReturnTypes.java:657]: you can use the precision to make this simpler, and perhaps safer too.
In some type systems maybe some of these types could be the same?

File: core/src/main/java/org/apache/calcite/sql/type/ReturnTypes.java (hunk 3 above)
Comment (normanj-bitquill) [reply core/src/main/java/org/apache/calcite/sql/type/ReturnTypes.java:638]: Since `NULL` is a possible result.
```
BITAND(1, NULL) = NULL
BITAND(CAST(x'01' AS BINARY(1)), NULL) = NULL
```

If nullability is set to `false` here, then `NULL` results are coerced back to integer types with the value `0`.

File: core/src/main/java/org/apache/calcite/sql/type/ReturnTypes.java (hunk 4 above)
Comment (normanj-bitquill) [reply core/src/main/java/org/apache/calcite/sql/type/ReturnTypes.java:645]: For line 642 above, it would get there when at least one argument is a non integer type and not null. This is probably not quite correct. I only want to exclude the `NULL` type. A nullable value is fine as long as it is an integer or binary type.

I'll see about correcting this.

File: core/src/main/java/org/apache/calcite/sql/type/ReturnTypes.java (hunk 5 above)
Comment (normanj-bitquill) [reply core/src/main/java/org/apache/calcite/sql/type/ReturnTypes.java]: Thanks, that helps a lot. I have cleaned up the code to make use of precision.

File: core/src/main/java/org/apache/calcite/sql/type/ReturnTypes.java (hunk 3 above)
Comment (normanj-bitquill) [reply core/src/main/java/org/apache/calcite/sql/type/ReturnTypes.java]: I see what you mean. I have reworked the return type detection. It is now only nullable if at least one argument type is nullable.

File: core/src/main/java/org/apache/calcite/sql/type/ReturnTypes.java (hunk 4 above)
Comment (normanj-bitquill) [reply core/src/main/java/org/apache/calcite/sql/type/ReturnTypes.java:649]: I have reworked the logic here.

If all arguments are of the NULL type, then the return type is INTEGER nullable.

If all arguments are integer types, then the return type is the largest type found. It is nullable if at least one argument is nullable.

As a fallback, the first argument that is not of the NULL type is the return type. It is nullable if at least one argument is nullable.

Comment (mihaibudiu) [reply core/src/main/java/org/apache/calcite/sql/type/ReturnTypes.java:649]: The fallback could be coded more explicitly as a check for the NULL type and then an assertion failure.
I like to be as explicit as possible, it's more future-proof.

Comment (mihaibudiu) [reply core/src/main/java/org/apache/calcite/sql/type/ReturnTypes.java:649]: In fact, if any argument has the NULL type the result type can also have the NULL type, since there is only one value with this type, NULL.

Comment (normanj-bitquill) [reply core/src/main/java/org/apache/calcite/sql/type/ReturnTypes.java:649]: I agree with you in theory, but I don't know how to do this in practice. When I make the change so that the NULL type is returned if any argument is of the NULL type, then the tests fail. They fail with this error occurring when trying to convert the result.
```
java.lang.NoSuchMethodException: java.lang.Void.valueOf(void)
```

It looks like it is trying to the result to a Void type in Java. It never went to an implementation of the `BITAND()` function. I also tried use a return type of `java.lang.Long` for the relevant implementation of the `BITAND()` function. This gives the a similar error but for the Long type error:
```
java.lang.NoSuchMethodException: java.lang.Long.voidValue()
```

Another way to look at this is that `BITAND(1, NULL)` should return a `NULL` value of the type of the first argument. I am treating `BITAND(1, NULL)` as the same as `BITAND(NULL, 1)`.

Comment (mihaibudiu) [reply core/src/main/java/org/apache/calcite/sql/type/ReturnTypes.java:649]: I was hoping that the nullpolicy would take care of that, but I guess I am wrong (the STRONG policy ensures that a function with null arguments is never called). But I guess you still need a return type, it can't be NULL. So probably the right solution is to return a compile-time error when all arguments have NULL type, and otherwise the type of the first non-NULL typed argument will do fine if any argument has type NULL.

Comment (normanj-bitquill) [reply core/src/main/java/org/apache/calcite/sql/type/ReturnTypes.java]: Updated. Now it gives an error when all arguments are NULL.

File: core/src/main/resources/org/apache/calcite/runtime/CalciteResource.properties [hunk 6]
```diff
@@ -207,6 +207,7 @@ ArgumentMustBeNumericLiteralInRange=Argument to function ''{0}'' must be a numer
 ValidationError=Validation Error: {0}
 IllegalLocaleFormat=Locale ''{0}'' in an illegal format
 ArgumentMustNotBeNull=Argument to function ''{0}'' must not be NULL
+AtLeastOneArgumentMustNotBeNull=At least one argument to function ''{0}'' must not be NULL
```
Comment (mihaibudiu) [inline core/src/main/resources/org/apache/calcite/runtime/CalciteResource.properties:210]: there are other functions which give an error in this case, but this error message is frankly better. maybe we should update the other ones.

**Review by mihaibudiu — APPROVED** [review]

**Comment by mihaibudiu** [comment]
Sorry, but it looks like other merges have introduced conflicts, please fix them so we can merge this one too

**Comment by normanj-bitquill** [comment]
Commits have been squashed.

File: core/src/main/resources/org/apache/calcite/runtime/CalciteResource.properties (hunk 6 above)
Comment (normanj-bitquill) [reply core/src/main/resources/org/apache/calcite/runtime/CalciteResource.properties:209]: Do you know what I can look for? I could update them.

Comment (mihaibudiu) [reply core/src/main/resources/org/apache/calcite/runtime/CalciteResource.properties:209]: I have seen this: "Illegal use of ''NULL''"

Comment (mihaibudiu) [reply core/src/main/resources/org/apache/calcite/runtime/CalciteResource.properties:209]: let's leave this for another PR, this has been strenuous enough
